kge/util/sampler.py

```diff
@@ -240,3 +241,51 @@ def _filter_and_resample_numba(
                         ctr += 1
                     num_found += len(idx)
                     num_remaining = num_new - num_found
+
+
+class KgeFrequencySampler(KgeSampler):
+    """
+    Sample negatives based on their relative occurrence in the slot in the train set.
+    Can be smoothed with a symmetric prior.
+    """
+    def __init__(self, config, configuration_key, dataset):
+        super().__init__(config, configuration_key, dataset)
+        self.samplers = []
+        alpha = self.get_option("symmetric_prior")
+        for slot in SLOTS:
+            vocab = np.arange(self.vocabulary_size[slot].item())
+            unique, counts = np.unique(
+                dataset._triples["train"][:, slot], return_counts=True
```

Comment:
Use ``dataset.train()``

Reply:
You can get the counts directly in the format you need using ``np.bincount(y, minlength=<vocab_size>)``.

Reply:
Then simply add the smoothing value.